{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### import the main functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/env python\n",
    "__author__ = 'heroico'\n",
    "\n",
    "import logging\n",
    "import numpy\n",
    "import os\n",
    "import gzip\n",
    "import ntpath\n",
    "import metax.WeightDBUtilities as WeightDBUtilities\n",
    "import metax.PrediXcanFormatUtilities as PrediXcanFormatUtilities\n",
    "import metax.ThousandGenomesUtilities as ThousandGenomesUtilities\n",
    "import metax.Logging as Logging\n",
    "import metax.Utilities as Utilities\n",
    "import metax.Formats as Formats\n",
    "\n",
    "\n",
    "def pathLeaf(path):\n",
    "    head, tail = ntpath.split(path)\n",
    "    return tail or ntpath.basename(head)\n",
    "\n",
    "class ProcessWeightDB(object):\n",
    "    def __init__(self, args):\n",
    "        self.weight_db = pathLeaf(args.weight_db)\n",
    "        self.db_path = args.weight_db\n",
    "        self.data_folder = args.input_folder\n",
    "        self.correlation_output = args.correlation_output\n",
    "        self.covariance_output = args.covariance_output\n",
    "        if args.covariance_output is None:\n",
    "            comp = os.path.splitext(self.weight_db)[0]\n",
    "            name = comp + \".cov.txt.gz\"\n",
    "            path = os.path.join(\"intermediate\", \"cov\")\n",
    "            path = os.path.join(path, name)\n",
    "            self.covariance_output = path\n",
    "\n",
    "        self.input_format = args.input_format\n",
    "\n",
    "        self.found_genes_for_covariance = {}\n",
    "        self.found_genes_for_correlation = {}\n",
    "\n",
    "        self.min_maf_filter = float(args.min_maf_filter) if args.min_maf_filter else None\n",
    "        self.max_maf_filter = float(args.max_maf_filter) if args.max_maf_filter else None\n",
    "\n",
    "        self.max_snps_in_gene = int(args.max_snps_in_gene) if args.max_snps_in_gene else None\n",
    "\n",
    "    def run(self):\n",
    "        if not self.correlation_output and not self.covariance_output:\n",
    "            logging.info(\"Provide --correlation_output or --covariance_output or both\")\n",
    "            return\n",
    "\n",
    "        logging.info(\"Loading Weights\")\n",
    "        weight_db_logic = WeightDBUtilities.WeightDBEntryLogic(self.db_path)\n",
    "\n",
    "        logging.info(\"Building files\")\n",
    "        self.buildFiles(weight_db_logic)\n",
    "\n",
    "        logging.info(\"Ran successfully\")\n",
    "\n",
    "    def buildFiles(self, weight_db_logic):\n",
    "        do_correlations = self.correlation_output is not None\n",
    "        if do_correlations:\n",
    "            if os.path.exists(self.correlation_output):\n",
    "                logging.info(\"%s already exists, delete it if you want it figured out again\", self.correlation_output)\n",
    "                do_correlations = False\n",
    "            else:\n",
    "                correlation_dir = os.path.dirname(self.correlation_output)\n",
    "                if not os.path.exists(correlation_dir):\n",
    "                    os.makedirs(correlation_dir)\n",
    "                self.writeFileHeader(self.correlation_output)\n",
    "\n",
    "        do_covariances = self.covariance_output is not None\n",
    "        if do_covariances:\n",
    "            if os.path.exists(self.covariance_output):\n",
    "                logging.info(\"%s already exists, delete it if you want it figured out again\", self.covariance_output)\n",
    "                do_covariances = False\n",
    "            else:\n",
    "                covariance_dir = os.path.dirname(self.covariance_output)\n",
    "                if not os.path.exists(covariance_dir):\n",
    "                    os.makedirs(covariance_dir)\n",
    "                self.writeFileHeader(self.covariance_output)\n",
    "\n",
    "        if not do_covariances and not do_correlations:\n",
    "            return\n",
    "\n",
    "        names = Utilities.dosageNamesFromFolder(self.data_folder)\n",
    "        for name in names:\n",
    "            snps, snps_by_rsid = self.getSNPS(name, weight_db_logic)\n",
    "            if do_correlations:\n",
    "                self.addToCorrelationFile(weight_db_logic, name, snps, snps_by_rsid)\n",
    "\n",
    "            if do_covariances:\n",
    "                self.addToCovarianceFile(weight_db_logic, name, snps, snps_by_rsid)\n",
    "\n",
    "    def writeFileHeader(self,path):\n",
    "        with gzip.open(path, \"ab\") as file:\n",
    "            file.write(\"GENE RSID1 RSID2 VALUE\\n\")\n",
    "\n",
    "    def getSNPS(self, name, weight_db_logic):\n",
    "        dosageLoader = None\n",
    "        if self.input_format == Formats.IMPUTE:\n",
    "            dosageLoader = ThousandGenomesUtilities.IMPUTEDosageLoader(self.data_folder, name) #outdated code\n",
    "        elif self.input_format == Formats.PrediXcan:\n",
    "            dosageName = Utilities.dosageName(name)\n",
    "            path = os.path.join(self.data_folder, dosageName)\n",
    "            dosageLoader = PrediXcanFormatUtilities.PrediXcanFormatDosageLoader(path, weight_db_logic)\n",
    "        else:\n",
    "            logging.info(\"Invalid input format: %s\", self.input_format)\n",
    "            return\n",
    "        snps, snps_by_rsid = dosageLoader.load()\n",
    "        return snps, snps_by_rsid\n",
    "\n",
    "    def addToCovarianceFile(self, weight_db_logic, name, snps, snps_by_rsid):\n",
    "        logging.info(\"Adding to covariance for %s-%s\", name, self.weight_db)\n",
    "\n",
    "        genes = weight_db_logic.weights_by_gene.keys()\n",
    "        total_genes = len(genes)\n",
    "        last_reported_percent = 0\n",
    "        processed = 0\n",
    "        for gene in genes:\n",
    "            processed += 1\n",
    "            percent = int(processed*100.0 / total_genes)\n",
    "            if percent == last_reported_percent+10:\n",
    "                logging.info(\"%d percent genes processed\", percent)\n",
    "                last_reported_percent = percent\n",
    "\n",
    "            entries = self.buildCovarianceEntries(name, gene, weight_db_logic, snps_by_rsid)\n",
    "\n",
    "            if len(entries) == 0:\n",
    "                logging.log(6,\"Gene %s has no snps in current file\", gene)\n",
    "                continue\n",
    "\n",
    "            self.addToFile(self.covariance_output, gene, entries)\n",
    "\n",
    "    def addToFile(self, path, gene, entries):\n",
    "        with gzip.open(path, \"ab\") as file:\n",
    "            for entry in entries:\n",
    "                line = \" \".join([gene, entry[0], entry[1], entry[2]])+\"\\n\"\n",
    "                file.write(line)\n",
    "\n",
    "    def buildCovarianceEntries(self, name, gene, weight_db_logic, snps_by_rsid):\n",
    "        weights_in_gene = weight_db_logic.weights_by_gene[gene]\n",
    "        rsids_from_genes = weights_in_gene.keys()\n",
    "\n",
    "        #gather as much data as we can work on\n",
    "        related_rsids, related_data = self.buildRelatedData(rsids_from_genes, snps_by_rsid, weights_in_gene)\n",
    "\n",
    "        if len(related_rsids) == 0:\n",
    "            return []\n",
    "\n",
    "        self.updateFoundCovariance(gene, name)\n",
    "\n",
    "        #covariance matrix of related SNP's data\n",
    "        array = numpy.array(related_data)\n",
    "        cov = numpy.cov(array)\n",
    "\n",
    "        #translate into sql entries\n",
    "        entries = self.buildMatrixOutputEntries(cov, rsids_from_genes, related_rsids, snps_by_rsid)\n",
    "        if not len(entries):\n",
    "            raise NameError(\"Couldn not build covariance entries for (%s,%s)\" %(name,gene))\n",
    "        return entries\n",
    "\n",
    "    def updateFoundCovariance(self, gene, name):\n",
    "        found = None\n",
    "        if gene in self.found_genes_for_covariance:\n",
    "            found = self.found_genes_for_covariance[gene]\n",
    "            logging.info(\"Gene %s found again for %s\", gene, name)\n",
    "        else:\n",
    "            found = []\n",
    "            self.found_genes_for_covariance[gene] = found\n",
    "        found.append(name)\n",
    "\n",
    "    def buildRelatedData(self, rsids_from_genes, snps_by_rsid, weights_in_gene):\n",
    "        related_rsids = []\n",
    "        related_data = []\n",
    "\n",
    "        l = len(rsids_from_genes)\n",
    "        if self.max_snps_in_gene and l > self.max_snps_in_gene:\n",
    "            logging.info(\"Skipping covariance too large: %d\", l)\n",
    "            return related_data, related_rsids\n",
    "\n",
    "        for rsid in rsids_from_genes:\n",
    "            if not rsid in snps_by_rsid:\n",
    "                logging.log(5, \"related rsid %s not present in genotype data\", rsid)\n",
    "                continue\n",
    "\n",
    "            related_snp = snps_by_rsid[rsid]\n",
    "            freq = sum(related_snp.data)*1.0/(2*len(related_snp.data))\n",
    "            if self.min_maf_filter and self.min_maf_filter > freq:\n",
    "                logging.log(6, \"related rsid %s below min maf: %s\", rsid, freq)\n",
    "                continue\n",
    "\n",
    "            if self.max_maf_filter and self.max_maf_filter < freq:\n",
    "                logging.log(6, \"related rsid %s  above max maf: %s\", rsid, freq)\n",
    "                continue\n",
    "\n",
    "            data = related_snp.data\n",
    "            weight = weights_in_gene[rsid]\n",
    "            if weight.ref_allele == related_snp.eff_allele and\\\n",
    "                weight.eff_allele == related_snp.ref_allele:\n",
    "                logging.log(7, \"related rsid %s has alleles flipped compared to model, transforming dosage\", rsid)\n",
    "                data = map(lambda x: 2-x, data)\n",
    "\n",
    "            related_data.append(data)\n",
    "            related_rsids.append(rsid)\n",
    "        return related_rsids, related_data\n",
    "\n",
    "    def buildMatrixOutputEntries(self, matrix, rsids_from_genes, related_rsids, snps_by_rsid):\n",
    "        entries = []\n",
    "\n",
    "        #special case: we might have a single rsid!\n",
    "        if matrix.ndim == 0:\n",
    "            c = str(float(matrix))\n",
    "            r = rsids_from_genes[0]\n",
    "            entries.append((r,r,c))\n",
    "            return entries\n",
    "\n",
    "        for i in xrange(0, len(rsids_from_genes)):\n",
    "            rsid_i = rsids_from_genes[i]\n",
    "            related_i = -1\n",
    "            if rsid_i in related_rsids:\n",
    "                related_i = related_rsids.index(rsid_i)\n",
    "\n",
    "            for j in xrange(0, len(rsids_from_genes)):\n",
    "                rsid_j = rsids_from_genes[j]\n",
    "\n",
    "                related_j = -1\n",
    "                if rsid_j in related_rsids:\n",
    "                    related_j = related_rsids.index(rsid_j)\n",
    "\n",
    "                value = \"NA\"\n",
    "                if related_i > -1 and related_j > -1:\n",
    "                    value = str(matrix[related_i][related_j])\n",
    "\n",
    "                if i == j:\n",
    "                    entries.append((rsid_i, rsid_i, value))\n",
    "                else:\n",
    "                    if value == \"NA\":\n",
    "                        if rsid_i < rsid_j:\n",
    "                            entries.append((rsid_i, rsid_j, value))\n",
    "                    else:\n",
    "                        snp_i = snps_by_rsid[rsid_i]\n",
    "                        snp_j = snps_by_rsid[rsid_j]\n",
    "\n",
    "                        if snp_i.position < snp_j.position:\n",
    "                            entries.append((rsid_i, rsid_j, value))\n",
    "        return entries\n",
    "\n",
    "    def addToCorrelationFile(self, weight_db_logic, name, snps, snps_by_rsid):\n",
    "        logging.info(\"Building correlation database for %s-%s\", name, self.weight_db)\n",
    "        genes = weight_db_logic.weights_by_gene.keys()\n",
    "        total_genes = len(genes)\n",
    "        last_reported_percent = 0\n",
    "        processed = 0\n",
    "        for gene in genes:\n",
    "            processed += 1\n",
    "            percent = int(processed*100.0 / total_genes)\n",
    "            if percent == last_reported_percent+10:\n",
    "                logging.info(\"%d percent genes processed\", percent)\n",
    "                last_reported_percent = percent\n",
    "\n",
    "            entries = self.buildCorrelationEntries(name, gene, weight_db_logic, snps_by_rsid)\n",
    "\n",
    "            if len(entries) == 0:\n",
    "                logging.log(6,\"Gene %s has no snps in current file\", gene)\n",
    "                continue\n",
    "\n",
    "            self.addToFile(self.correlation_output, gene, entries)\n",
    "\n",
    "    def buildCorrelationEntries(self, name, gene, weight_db_logic, snps_by_rsid):\n",
    "        weights_in_gene = weight_db_logic.weights_by_gene[gene]\n",
    "        rsids_from_genes = weights_in_gene.keys()\n",
    "\n",
    "        #gather as much data as we can work on\n",
    "        related_rsids, related_data = self.buildRelatedData(rsids_from_genes, snps_by_rsid, weights_in_gene)\n",
    "\n",
    "        if len(related_rsids) == 0:\n",
    "            return []\n",
    "\n",
    "        self.updateFoundCorrelation(gene, name)\n",
    "\n",
    "        #correlation matrix of related SNP's data\n",
    "        array = numpy.array(related_data)\n",
    "        cor = numpy.corrcoef(array)\n",
    "\n",
    "        #translate into sql entries\n",
    "        entries = self.buildMatrixOutputEntries(cor, rsids_from_genes, related_rsids, snps_by_rsid)\n",
    "        if not len(entries):\n",
    "            raise NameError(\"Couldn not build correlation entries for (%s,%s)\" %(name,gene))\n",
    "        return entries\n",
    "\n",
    "    def updateFoundCorrelation(self, gene, name):\n",
    "        found = None\n",
    "        if gene in self.found_genes_for_correlation:\n",
    "            found = self.found_genes_for_correlation[gene]\n",
    "            logging.info(\"Gene %s found again for %s\", gene, name)\n",
    "        else:\n",
    "            found = []\n",
    "            self.found_genes_for_correlation[gene] = found\n",
    "        found.append(name)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### test argument\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### R interface\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import rpy2\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from rpy2.robjects import r\n",
    "import rpy2.robjects.packages as rpackages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "rpy2.rinterface.NULL"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "utils = rpackages.importr('utils')\n",
    "utils.chooseCRANmirror(ind=1) # select the first mirror in the list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: also installing the dependencies ‘Rcpp’, ‘mvtnorm’, ‘SKAT’, ‘BH’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: trying URL 'https://cloud.r-project.org/src/contrib/Rcpp_0.12.14.tar.gz'\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: Content type 'application/x-gzip'\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning:  length 3756532 bytes (3.6 MB)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: =\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: \n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: downloaded 3.6 MB\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: trying URL 'https://cloud.r-project.org/src/contrib/mvtnorm_1.0-6.tar.gz'\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning:  length 159232 bytes (155 KB)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: downloaded 155 KB\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: trying URL 'https://cloud.r-project.org/src/contrib/SKAT_1.3.2.1.tar.gz'\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning:  length 1162383 bytes (1.1 MB)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: downloaded 1.1 MB\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: trying URL 'https://cloud.r-project.org/src/contrib/BH_1.65.0-1.tar.gz'\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning:  length 10839577 bytes (10.3 MB)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: downloaded 10.3 MB\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: trying URL 'https://cloud.r-project.org/src/contrib/GBJ_0.5.0.tar.gz'\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning:  length 48612 bytes (47 KB)\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: downloaded 47 KB\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: \n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: The downloaded source packages are in\n",
      "\t‘/private/var/folders/_4/6kyw9fpj2lj83l22xwmfxx6w0000gn/T/RtmprVUnkI/downloaded_packages’\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: Updating HTML index of packages in '.Library'\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: Making 'packages.html' ...\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning:  done\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "# R package names\n",
    "names_to_install = ['GBJ']\n",
    "\n",
    "# R vector of strings\n",
    "from rpy2.robjects.vectors import StrVector\n",
    "\n",
    "# Selectively install what needs to be install.\n",
    "# We are fancy, just because we can.\n",
    "\n",
    "if len(names_to_install) > 0:\n",
    "    utils.install_packages(StrVector(names_to_install))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: Error in loadNamespace(name) : there is no package called ‘GBJ’\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: In addition: \n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: Warning messages:\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: 1: \n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: In (function (pkgs, lib, repos = getOption(\"repos\"), contriburl = contrib.url(repos,  :\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: \n",
      " \n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning:  installation of package ‘Rcpp’ had non-zero exit status\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: 2: \n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning:  installation of package ‘mvtnorm’ had non-zero exit status\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: 3: \n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning:  installation of package ‘SKAT’ had non-zero exit status\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: 4: \n",
      "  warnings.warn(x, RRuntimeWarning)\n",
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning:  installation of package ‘GBJ’ had non-zero exit status\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n"
     ]
    },
    {
     "ename": "RRuntimeError",
     "evalue": "Error in loadNamespace(name) : there is no package called ‘GBJ’\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRRuntimeError\u001b[0m                             Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-13-360e9646739d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mimportr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"GBJ\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/robjects/packages.pyc\u001b[0m in \u001b[0;36mimportr\u001b[0;34m(name, lib_loc, robject_translations, signature_translation, suppress_messages, on_conflict, symbol_r2python, symbol_check_after, data)\u001b[0m\n\u001b[1;32m    451\u001b[0m     if _package_has_namespace(rname, \n\u001b[1;32m    452\u001b[0m                               _system_file(package = rname)):\n\u001b[0;32m--> 453\u001b[0;31m         \u001b[0menv\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_get_namespace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    454\u001b[0m         \u001b[0mversion\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_get_namespace_version\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    455\u001b[0m         \u001b[0mexported_names\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_get_namespace_exports\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mRRuntimeError\u001b[0m: Error in loadNamespace(name) : there is no package called ‘GBJ’\n"
     ]
    }
   ],
   "source": [
    "importr(\"GBJ\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import sqlite3\n",
    "\n",
    "\n",
    "# function of creating connection\n",
    "def create_connection(db_file):\n",
    "    \"\"\" create a database connection to the SQLite database\n",
    "        specified by the db_file\n",
    "    :param db_file: database file\n",
    "    :return: Connection object or None\n",
    "    \"\"\"\n",
    "    try:\n",
    "        conn = sqlite3.connect(db_file)\n",
    "        return conn\n",
    "    except sqlite3.Error as e:\n",
    "        print(e)\n",
    "        sys.exit(1)\n",
    "    return None\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "dbname = \"/Users/jerome/Projects/metaxcan/MetaXcan/v1/data/Stomach.db\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "conn = create_connection(dbname)\n",
    "cur = conn.cursor() \n",
    "gene = '\"A3GALT2\"'\n",
    "sql = \"select * from weights where gene =\" + gene\n",
    "tmp_query = cur.execute(sql).fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'select * from weights where gene =\"A3GALT2\"'"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sql\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(u'rs644879', u'A3GALT2', 6.27340257627e-05, u'T', u'C'),\n",
       " (u'rs6671539', u'A3GALT2', -0.000263706411293, u'G', u'A'),\n",
       " (u'rs34665119', u'A3GALT2', 0.00136937184868, u'A', u'G'),\n",
       " (u'rs11802938', u'A3GALT2', 0.000224462896528, u'G', u'A'),\n",
       " (u'rs1566230', u'A3GALT2', 0.000213158781561, u'T', u'C'),\n",
       " (u'rs12040608', u'A3GALT2', 2.0053067055e-05, u'C', u'T'),\n",
       " (u'rs12040669', u'A3GALT2', 0.00213220904559, u'C', u'T'),\n",
       " (u'rs10914671', u'A3GALT2', -0.000780960140634, u'C', u'T'),\n",
       " (u'rs6670552', u'A3GALT2', -0.000323079018941, u'C', u'T'),\n",
       " (u'rs12751707', u'A3GALT2', 0.000545219903371, u'A', u'G'),\n",
       " (u'rs4077439', u'A3GALT2', -0.00177721129844, u'G', u'A'),\n",
       " (u'rs12736971', u'A3GALT2', 0.003386056752, u'A', u'G'),\n",
       " (u'rs12567237', u'A3GALT2', 0.000969534243751, u'T', u'C'),\n",
       " (u'rs28880434', u'A3GALT2', 0.000673745385822, u'C', u'T'),\n",
       " (u'rs12726791', u'A3GALT2', 0.000642417574073, u'C', u'T'),\n",
       " (u'rs12723104', u'A3GALT2', 0.00112388295962, u'T', u'C'),\n",
       " (u'rs4641276', u'A3GALT2', -0.00146576563367, u'T', u'C'),\n",
       " (u'rs7527638', u'A3GALT2', 3.7845728255e-05, u'C', u'T'),\n",
       " (u'rs12041371', u'A3GALT2', 2.94000867471e-05, u'T', u'C'),\n",
       " (u'rs7523119', u'A3GALT2', -9.46718410663e-05, u'G', u'A'),\n",
       " (u'rs10798937', u'A3GALT2', 0.00199879978827, u'G', u'A'),\n",
       " (u'rs12734172', u'A3GALT2', 0.00366719426495, u'G', u'A'),\n",
       " (u'rs11061', u'A3GALT2', 0.00245416293977, u'G', u'A'),\n",
       " (u'rs11554674', u'A3GALT2', 0.00599062650516, u'G', u'A'),\n",
       " (u'rs4653044', u'A3GALT2', -0.00011686167905, u'A', u'C'),\n",
       " (u'rs12127865', u'A3GALT2', -0.000494935927443, u'A', u'G'),\n",
       " (u'rs6673519', u'A3GALT2', 0.00129788594882, u'C', u'T'),\n",
       " (u'rs12041636', u'A3GALT2', 1.01097486252e-05, u'T', u'C'),\n",
       " (u'rs12727428', u'A3GALT2', 8.49330277311e-05, u'G', u'A'),\n",
       " (u'rs11577204', u'A3GALT2', 0.00211817652866, u'G', u'A'),\n",
       " (u'rs6699152', u'A3GALT2', 0.000121410788492, u'G', u'A'),\n",
       " (u'rs12033491', u'A3GALT2', 2.25545015948e-05, u'A', u'G'),\n",
       " (u'rs4357495', u'A3GALT2', 0.000794405442648, u'G', u'A'),\n",
       " (u'rs10914716', u'A3GALT2', 7.10710129275e-06, u'C', u'T'),\n",
       " (u'rs1570686', u'A3GALT2', 5.63023844759e-06, u'G', u'A'),\n",
       " (u'rs12032426', u'A3GALT2', 0.00089950710113, u'C', u'A'),\n",
       " (u'rs12022060', u'A3GALT2', 0.000151050032475, u'A', u'G'),\n",
       " (u'rs12025675', u'A3GALT2', 0.00266018995625, u'G', u'A'),\n",
       " (u'rs10914726', u'A3GALT2', 0.0019532627603, u'G', u'A'),\n",
       " (u'rs114013172', u'A3GALT2', 0.00341271663275, u'A', u'G'),\n",
       " (u'rs12737304', u'A3GALT2', -1.57832044879e-06, u'C', u'T'),\n",
       " (u'rs1417964', u'A3GALT2', -0.0007598466269, u'T', u'C'),\n",
       " (u'rs2483842', u'A3GALT2', -4.68077009676e-05, u'A', u'G'),\n",
       " (u'rs4350172', u'A3GALT2', -2.92968462112e-06, u'C', u'A'),\n",
       " (u'rs477438', u'A3GALT2', 0.000748091886105, u'A', u'G'),\n",
       " (u'rs4652974', u'A3GALT2', -0.000181651625495, u'T', u'C'),\n",
       " (u'rs507641', u'A3GALT2', 0.000226613557346, u'T', u'G'),\n",
       " (u'rs472363', u'A3GALT2', 9.9518455555e-06, u'T', u'C'),\n",
       " (u'rs508927', u'A3GALT2', 5.94436093949e-05, u'T', u'C'),\n",
       " (u'rs549048', u'A3GALT2', 0.000224412333589, u'T', u'C'),\n",
       " (u'rs771131', u'A3GALT2', 8.45030746577e-06, u'T', u'G')]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp_query\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "ename": "OperationalError",
     "evalue": "no such column: ref",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mOperationalError\u001b[0m                          Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-135-3cac7d502018>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0msql\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"select rsid, ref from weights where gene =\"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mgene\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mtmp_query\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcur\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexecute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msql\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfetchall\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mOperationalError\u001b[0m: no such column: ref"
     ]
    }
   ],
   "source": [
    "sql = \"select rsid, ref from weights where gene =\" + gene\n",
    "\n",
    "tmp_query = cur.execute(sql).fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(0, u'rsid', u'TEXT', 0, None, 0),\n",
       " (1, u'gene', u'TEXT', 0, None, 0),\n",
       " (2, u'weight', u'DOUBLE', 0, None, 0),\n",
       " (3, u'ref_allele', u'CHARACTER', 0, None, 0),\n",
       " (4, u'eff_allele', u'CHARACTER', 0, None, 0)]"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cur.execute(\"PRAGMA table_info(weights)\").fetchall()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "str(tmp_query[0])\n",
    "tmp_rsid = map(lambda x: str(x[0]), tmp_query)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['rs644879',\n",
       " 'rs6671539',\n",
       " 'rs34665119',\n",
       " 'rs11802938',\n",
       " 'rs1566230',\n",
       " 'rs12040608',\n",
       " 'rs12040669',\n",
       " 'rs10914671',\n",
       " 'rs6670552',\n",
       " 'rs12751707',\n",
       " 'rs4077439',\n",
       " 'rs12736971',\n",
       " 'rs12567237',\n",
       " 'rs28880434',\n",
       " 'rs12726791',\n",
       " 'rs12723104',\n",
       " 'rs4641276',\n",
       " 'rs7527638',\n",
       " 'rs12041371',\n",
       " 'rs7523119',\n",
       " 'rs10798937',\n",
       " 'rs12734172',\n",
       " 'rs11061',\n",
       " 'rs11554674',\n",
       " 'rs4653044',\n",
       " 'rs12127865',\n",
       " 'rs6673519',\n",
       " 'rs12041636',\n",
       " 'rs12727428',\n",
       " 'rs11577204',\n",
       " 'rs6699152',\n",
       " 'rs12033491',\n",
       " 'rs4357495',\n",
       " 'rs10914716',\n",
       " 'rs1570686',\n",
       " 'rs12032426',\n",
       " 'rs12022060',\n",
       " 'rs12025675',\n",
       " 'rs10914726',\n",
       " 'rs114013172',\n",
       " 'rs12737304',\n",
       " 'rs1417964',\n",
       " 'rs2483842',\n",
       " 'rs4350172',\n",
       " 'rs477438',\n",
       " 'rs4652974',\n",
       " 'rs507641',\n",
       " 'rs472363',\n",
       " 'rs508927',\n",
       " 'rs549048',\n",
       " 'rs771131']"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp_rsid\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/ipykernel_launcher.py:1: FutureWarning: pd.match() is deprecated and will be removed in a future version\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n"
     ]
    }
   ],
   "source": [
    "index = pd.match(tmp_rsid, tmp_rsid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/ipykernel_launcher.py:1: FutureWarning: pd.match() is deprecated and will be removed in a future version\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n"
     ]
    }
   ],
   "source": [
    "tmp_array = pd.match([1,2,3,9],[2,3,5,7])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "numpy.ndarray"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(tmp_array)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 1])"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp_array[tmp_array>-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([False,  True,  True, False], dtype=bool)"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp_array > -1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "only integer scalar arrays can be converted to a scalar index",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-55-4683ee07ae66>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mtmp_rsid\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mtmp_array\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m: only integer scalar arrays can be converted to a scalar index"
     ]
    }
   ],
   "source": [
    "tmp_rsid[tmp_array]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-1,  0])"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp_array[tmp_array[tmp_array>-1].tolist()]\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 1])"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp_array[tmp_array>-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['rs6671539', 'rs34665119', 'rs11802938'],\n",
       "      dtype='|S11')"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.array(tmp_rsid)[np.array([1,2,3])]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([7, 2, 7, 4, 7])"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.array(tmp_rsid)[:3]\n",
    "c = np.array([1,2,3,4,5])\n",
    "c[np.array([0,2,4])] = 7,7,7\n",
    "c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/ipykernel_launcher.py:3: FutureWarning: pd.match() is deprecated and will be removed in a future version\n",
      "  This is separate from the ipykernel package so we can avoid doing imports until\n"
     ]
    }
   ],
   "source": [
    "a = np.array([1,2,3,4,5])\n",
    "b = np.array([9,1,2,3,4,5,6,7,8])\n",
    "indexab = pd.match(a,b)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1, 2, 3, 4, 5])"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indexab\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ True,  True,  True,  True,  True], dtype=bool)"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indexab > 0\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1, 2, 3, 4, 5])"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a[indexab > 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#joint GBJ\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cov_matrix = np.loadtxt(\"/Users/jerome/Projects/metaxcan/MetaXcan/v1/data/TOP3B.cov\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(94, 94)"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cov_matrix.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "snp_rsid = pd.read_table(\"/Users/jerome/Projects/metaxcan/MetaXcan/v1/data/TOP3B.snplist\",header =None)\n",
    "snp_rsid = list(snp_rsid.loc[:,0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/ipykernel_launcher.py:13: FutureWarning: pd.match() is deprecated and will be removed in a future version\n",
      "  del sys.path[0]\n"
     ]
    }
   ],
   "source": [
    "M = len(snp_rsid) #number of snps\n",
    "N = 1\n",
    "gene = \"TOP3B\"\n",
    "weights = np.zeros(shape = (M, N))\n",
    "#for i in range(N):\n",
    "i = 0\n",
    "dbname = \"/Users/jerome/Projects/metaxcan/MetaXcan/v1/data/Stomach.db\"\n",
    "conn = create_connection(dbname)\n",
    "cur = conn.cursor()  \n",
    "sql_q = 'select * from weights where gene = \"' + gene + '\"'\n",
    "tmp_query = cur.execute(sql_q).fetchall()\n",
    "rsid_in_db = map(lambda x: str(x[0]), tmp_query)\n",
    "index = pd.match(rsid_in_db, snp_rsid)\n",
    "indi = index[index > -1]\n",
    "# extract the weight\n",
    "sql_q = 'select * from weights where gene = \"' + gene + '\"'\n",
    "tmp_query = cur.execute(sql_q).fetchall()\n",
    "tmp_weights = np.array(map(lambda x: float(x[2]), tmp_query))\n",
    "weights[indi,i] = tmp_weights[index > -1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'select * from weights where gene = \"TOP3B\"'"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sql_q"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([  5.95338282e-04,   8.96832937e-06,  -5.15409039e-05,\n",
       "         3.07426030e-03,   2.27576594e-04,   5.65074472e-03,\n",
       "        -4.96734167e-03,   1.56079960e-03,  -1.48322517e-03,\n",
       "        -2.45901074e-04,  -6.13580956e-05,  -7.19388993e-03,\n",
       "         1.24588148e-03,   4.22629443e-03,   4.07837805e-04,\n",
       "         3.03761490e-03,  -1.17454223e-03,  -7.91966346e-04,\n",
       "        -2.02634474e-04,   5.16420029e-03,  -1.22498382e-03,\n",
       "         1.34826208e-02,   3.95381306e-03,  -1.69581402e-02,\n",
       "        -1.93481755e-05,  -3.15025361e-04,  -5.27540909e-04,\n",
       "        -3.08823855e-03,  -8.32554991e-04,  -3.51225620e-03,\n",
       "        -6.85744264e-05,  -1.44968372e-02,  -5.64451716e-03,\n",
       "        -3.05647848e-03,  -1.20068994e-04,  -1.52938942e-02,\n",
       "        -1.68395398e-03,  -3.48585118e-03,  -2.10373642e-03,\n",
       "        -4.77514660e-03,  -1.27122491e-02,  -5.89674135e-02,\n",
       "        -2.65188078e-03,  -2.78987295e-03,  -1.44375498e-02,\n",
       "        -6.02239092e-03,  -7.75847601e-03,  -2.37549158e-02,\n",
       "        -1.14257057e-03,  -8.04822266e-03,  -2.39023675e-02,\n",
       "        -6.96311821e-03,  -2.54682455e-03,  -5.18245046e-03,\n",
       "        -2.20904632e-02,  -9.03450528e-03,  -1.02129407e-02,\n",
       "        -1.21523565e-02,  -8.12975278e-03,  -4.35622686e-03,\n",
       "        -1.85068864e-02,  -1.53602779e-02,  -4.67824169e-03,\n",
       "        -3.49675911e-02,  -1.34850193e-02,  -2.24636899e-02,\n",
       "        -2.64026795e-02,  -1.04272965e-02,  -1.84150380e-02,\n",
       "        -3.40208026e-03,   2.24152899e-03,   6.18335567e-03,\n",
       "         3.90368092e-03,   9.51424144e-04,   1.15161066e-03,\n",
       "        -7.27498126e-03,  -2.77683146e-05,  -6.66732128e-05,\n",
       "        -2.08687512e-04,  -1.38312469e-03,  -1.62197066e-03,\n",
       "        -1.69497148e-04,  -2.29821392e-03,  -1.05019318e-04,\n",
       "        -4.73058386e-04,   3.12192535e-04,   4.82302711e-06,\n",
       "         9.78015048e-04,  -2.54057880e-04,   3.05410230e-04,\n",
       "         6.42718596e-05,  -1.09079159e-04,  -1.01795860e-03,\n",
       "        -2.44075018e-04])"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp_weights"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['rs35607949',\n",
       " 'rs8141374',\n",
       " 'rs1005579',\n",
       " 'rs9610631',\n",
       " 'rs2266965',\n",
       " 'rs1988315',\n",
       " 'rs2266966',\n",
       " 'rs5995257',\n",
       " 'rs4822783',\n",
       " 'rs389624',\n",
       " 'rs11490038',\n",
       " 'rs10427625',\n",
       " 'rs79217851',\n",
       " 'rs2330019',\n",
       " 'rs2330018',\n",
       " 'rs4145537',\n",
       " 'rs2541937',\n",
       " 'rs2266979',\n",
       " 'rs2266977',\n",
       " 'rs449597',\n",
       " 'rs2283797',\n",
       " 'rs9610704',\n",
       " 'rs5756323',\n",
       " 'rs11089937',\n",
       " 'rs151680',\n",
       " 'rs9620740',\n",
       " 'rs5997304',\n",
       " 'rs410356',\n",
       " 'rs743517',\n",
       " 'rs743516',\n",
       " 'rs9610794',\n",
       " 'rs8139376',\n",
       " 'rs9610796',\n",
       " 'rs2329884',\n",
       " 'rs11089841',\n",
       " 'rs867117',\n",
       " 'rs743526',\n",
       " 'rs4821629',\n",
       " 'rs238775',\n",
       " 'rs12158712',\n",
       " 'rs9607413',\n",
       " 'rs56406117',\n",
       " 'rs4820285',\n",
       " 'rs9610769',\n",
       " 'rs8141816',\n",
       " 'rs2083565',\n",
       " 'rs6005623',\n",
       " 'rs2027789',\n",
       " 'rs239915',\n",
       " 'rs239914',\n",
       " 'rs9306459',\n",
       " 'rs5750831',\n",
       " 'rs9607401',\n",
       " 'rs450129',\n",
       " 'rs9607241',\n",
       " 'rs56021049',\n",
       " 'rs62234965',\n",
       " 'rs1669124',\n",
       " 'rs9625334',\n",
       " 'rs5750426',\n",
       " 'rs5750553',\n",
       " 'rs5750552',\n",
       " 'rs2266980',\n",
       " 'rs5750807',\n",
       " 'rs369463',\n",
       " 'rs6001231',\n",
       " 'rs9607473',\n",
       " 'rs6001121',\n",
       " 'rs882097',\n",
       " 'rs9611215',\n",
       " 'rs861787',\n",
       " 'rs894095',\n",
       " 'rs986643',\n",
       " 'rs986645',\n",
       " 'rs1669116',\n",
       " 'rs240068',\n",
       " 'rs9607460',\n",
       " 'rs9607461',\n",
       " 'rs9610750',\n",
       " 'rs2329929',\n",
       " 'rs9610690',\n",
       " 'rs56094396',\n",
       " 'rs7349039',\n",
       " 'rs5994846',\n",
       " 'rs5756985',\n",
       " 'rs381523',\n",
       " 'rs9610317',\n",
       " 'rs861812',\n",
       " 'rs9610608',\n",
       " 'rs13340063',\n",
       " 'rs383737',\n",
       " 'rs1210397',\n",
       " 'rs9610728',\n",
       " 'rs6001413']"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_rsid\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[ -2.44075018e-04],\n",
       "       [ -1.22498382e-03],\n",
       "       [ -6.02239092e-03],\n",
       "       [ -1.68395398e-03],\n",
       "       [ -3.15025361e-04],\n",
       "       [ -1.09079159e-04],\n",
       "       [ -5.27540909e-04],\n",
       "       [ -1.27122491e-02],\n",
       "       [  5.95338282e-04],\n",
       "       [ -5.15409039e-05],\n",
       "       [ -8.32554991e-04],\n",
       "       [ -1.34850193e-02],\n",
       "       [ -6.85744264e-05],\n",
       "       [  9.78015048e-04],\n",
       "       [  4.82302711e-06],\n",
       "       [ -1.05019318e-04],\n",
       "       [  5.65074472e-03],\n",
       "       [ -4.67824169e-03],\n",
       "       [ -1.85068864e-02],\n",
       "       [  1.24588148e-03],\n",
       "       [ -2.39023675e-02],\n",
       "       [ -1.14257057e-03],\n",
       "       [ -1.04272965e-02],\n",
       "       [ -1.62197066e-03],\n",
       "       [ -3.05647848e-03],\n",
       "       [  2.27576594e-04],\n",
       "       [ -1.48322517e-03],\n",
       "       [  1.56079960e-03],\n",
       "       [ -2.29821392e-03],\n",
       "       [ -1.69497148e-04],\n",
       "       [ -2.24636899e-02],\n",
       "       [ -1.93481755e-05],\n",
       "       [ -2.64026795e-02],\n",
       "       [ -3.48585118e-03],\n",
       "       [ -1.21523565e-02],\n",
       "       [ -8.04822266e-03],\n",
       "       [ -8.12975278e-03],\n",
       "       [  2.24152899e-03],\n",
       "       [ -3.40208026e-03],\n",
       "       [ -3.51225620e-03],\n",
       "       [ -2.78987295e-03],\n",
       "       [ -1.53602779e-02],\n",
       "       [  1.15161066e-03],\n",
       "       [ -4.35622686e-03],\n",
       "       [ -4.96734167e-03],\n",
       "       [ -4.77514660e-03],\n",
       "       [ -2.45901074e-04],\n",
       "       [ -5.89674135e-02],\n",
       "       [ -6.96311821e-03],\n",
       "       [ -2.54682455e-03],\n",
       "       [ -6.13580956e-05],\n",
       "       [  6.42718596e-05],\n",
       "       [ -1.20068994e-04],\n",
       "       [  4.22629443e-03],\n",
       "       [ -1.69581402e-02],\n",
       "       [ -5.64451716e-03],\n",
       "       [ -1.44375498e-02],\n",
       "       [ -1.17454223e-03],\n",
       "       [  3.07426030e-03],\n",
       "       [  6.18335567e-03],\n",
       "       [ -6.66732128e-05],\n",
       "       [ -2.77683146e-05],\n",
       "       [ -3.49675911e-02],\n",
       "       [  3.05410230e-04],\n",
       "       [  3.03761490e-03],\n",
       "       [ -4.73058386e-04],\n",
       "       [ -1.84150380e-02],\n",
       "       [ -1.38312469e-03],\n",
       "       [ -1.01795860e-03],\n",
       "       [  3.12192535e-04],\n",
       "       [ -7.19388993e-03],\n",
       "       [ -2.10373642e-03],\n",
       "       [  3.90368092e-03],\n",
       "       [  9.51424144e-04],\n",
       "       [ -7.91966346e-04],\n",
       "       [ -2.65188078e-03],\n",
       "       [ -9.03450528e-03],\n",
       "       [ -1.02129407e-02],\n",
       "       [ -2.20904632e-02],\n",
       "       [ -7.27498126e-03],\n",
       "       [ -7.75847601e-03],\n",
       "       [ -1.52938942e-02],\n",
       "       [  5.16420029e-03],\n",
       "       [  1.34826208e-02],\n",
       "       [ -2.08687512e-04],\n",
       "       [ -2.37549158e-02],\n",
       "       [ -3.08823855e-03],\n",
       "       [ -2.02634474e-04],\n",
       "       [ -1.44968372e-02],\n",
       "       [  3.95381306e-03],\n",
       "       [  8.96832937e-06],\n",
       "       [  4.07837805e-04],\n",
       "       [ -5.18245046e-03],\n",
       "       [ -2.54057880e-04]])"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "weights\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cov_gene = weights.T * cov_matrix * weights"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([  2.91712493e-08,   2.32080179e-10,   1.34355786e-08,\n",
       "        -1.35915292e-09,   1.23629803e-10,   1.30368743e-08])"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cov_gene.shape\n",
    "cov_gene[0,:6]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for i in range(N):\n",
    "    if cov_gene[i,i] != 0:\n",
    "        cov_gene[i,:] = cov_gene[i,:] / np.sqrt(cov_gene[i,i])\n",
    "        cov_gene[:,i] = cov_gene[:,i] / cov_gene[i,i]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([  1.00000000e+00,   1.35881564e-06,   7.86645134e-05,\n",
       "        -7.95775946e-06,   7.23845142e-07,   7.63301234e-05])"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cov_gene[0,:6]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### mask file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "zscore_dict = {}\n",
    "i = 0\n",
    "nam = \"zscore_\" + str(i+1)\n",
    "zscore_dict[nam] = pd.read_csv(\"/Users/jerome/Projects/metaxcan/MetaXcan/v1/data/Stomach.csv\", header = \"infer\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>zscore</th>\n",
       "      <th>effect_size</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>var_g</th>\n",
       "      <th>pred_perf_r2</th>\n",
       "      <th>pred_perf_pval</th>\n",
       "      <th>pred_perf_qval</th>\n",
       "      <th>n_snps_used</th>\n",
       "      <th>n_snps_in_cov</th>\n",
       "      <th>n_snps_in_model</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CEACAM19</td>\n",
       "      <td>CEACAM19</td>\n",
       "      <td>7.570289</td>\n",
       "      <td>0.494302</td>\n",
       "      <td>3.723957e-14</td>\n",
       "      <td>3.791226e-02</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>143</td>\n",
       "      <td>162.0</td>\n",
       "      <td>163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EXTL3</td>\n",
       "      <td>EXTL3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-23.451963</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.264877e-05</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>4</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CLPTM1</td>\n",
       "      <td>CLPTM1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.099350</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.464133e-03</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>21</td>\n",
       "      <td>22.0</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CLU</td>\n",
       "      <td>CLU</td>\n",
       "      <td>6.973488</td>\n",
       "      <td>0.043537</td>\n",
       "      <td>3.091788e-12</td>\n",
       "      <td>3.343004e+00</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>21</td>\n",
       "      <td>21.0</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BCAM</td>\n",
       "      <td>BCAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.538550</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.117730e-04</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PPP1R37</td>\n",
       "      <td>PPP1R37</td>\n",
       "      <td>6.430535</td>\n",
       "      <td>15.715981</td>\n",
       "      <td>1.271553e-10</td>\n",
       "      <td>3.665028e-05</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>5</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>EPHA1</td>\n",
       "      <td>EPHA1</td>\n",
       "      <td>6.103094</td>\n",
       "      <td>1.449575</td>\n",
       "      <td>1.040343e-09</td>\n",
       "      <td>2.392429e-03</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>26</td>\n",
       "      <td>27.0</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CASP2</td>\n",
       "      <td>CASP2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.117632</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.083994e-04</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>17</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MTCH2</td>\n",
       "      <td>MTCH2</td>\n",
       "      <td>-5.462552</td>\n",
       "      <td>-2.070170</td>\n",
       "      <td>4.693376e-08</td>\n",
       "      <td>9.155414e-04</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>36</td>\n",
       "      <td>37.0</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CD2AP</td>\n",
       "      <td>CD2AP</td>\n",
       "      <td>4.860381</td>\n",
       "      <td>0.291114</td>\n",
       "      <td>1.171603e-06</td>\n",
       "      <td>3.277740e-02</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>37</td>\n",
       "      <td>37.0</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SYTL2</td>\n",
       "      <td>SYTL2</td>\n",
       "      <td>-4.846171</td>\n",
       "      <td>-0.574747</td>\n",
       "      <td>1.258672e-06</td>\n",
       "      <td>8.906129e-03</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>36</td>\n",
       "      <td>38.0</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>FAM180B</td>\n",
       "      <td>FAM180B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-24.396055</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.581500e-06</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PTGDR2</td>\n",
       "      <td>PTGDR2</td>\n",
       "      <td>-4.586619</td>\n",
       "      <td>-1.197283</td>\n",
       "      <td>4.504807e-06</td>\n",
       "      <td>1.779324e-03</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>12</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NDUFS3</td>\n",
       "      <td>NDUFS3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14.895390</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.271488e-05</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>WDR55</td>\n",
       "      <td>WDR55</td>\n",
       "      <td>4.464619</td>\n",
       "      <td>1.377777</td>\n",
       "      <td>8.021139e-06</td>\n",
       "      <td>1.333930e-03</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>55</td>\n",
       "      <td>60.0</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>AIMP2</td>\n",
       "      <td>AIMP2</td>\n",
       "      <td>-4.463330</td>\n",
       "      <td>-2.005129</td>\n",
       "      <td>8.069562e-06</td>\n",
       "      <td>1.141239e-03</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>17</td>\n",
       "      <td>22.0</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CD14</td>\n",
       "      <td>CD14</td>\n",
       "      <td>4.414933</td>\n",
       "      <td>0.492409</td>\n",
       "      <td>1.010413e-05</td>\n",
       "      <td>5.269745e-03</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>18</td>\n",
       "      <td>19.0</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ZP1</td>\n",
       "      <td>ZP1</td>\n",
       "      <td>-4.382206</td>\n",
       "      <td>-2.423554</td>\n",
       "      <td>1.174834e-05</td>\n",
       "      <td>4.078626e-04</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>43</td>\n",
       "      <td>47.0</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MAFK</td>\n",
       "      <td>MAFK</td>\n",
       "      <td>4.339047</td>\n",
       "      <td>0.301414</td>\n",
       "      <td>1.431018e-05</td>\n",
       "      <td>2.929442e-02</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>34</td>\n",
       "      <td>47.0</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>RAPSN</td>\n",
       "      <td>RAPSN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-83.548762</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.574258e-07</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>2</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>HBEGF</td>\n",
       "      <td>HBEGF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-9.826707</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.437816e-05</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>3</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>DMWD</td>\n",
       "      <td>DMWD</td>\n",
       "      <td>4.148808</td>\n",
       "      <td>0.431981</td>\n",
       "      <td>3.342103e-05</td>\n",
       "      <td>1.295833e-02</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>55</td>\n",
       "      <td>57.0</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>MDK</td>\n",
       "      <td>MDK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.963345</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.461208e-03</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>4</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>POLR2E</td>\n",
       "      <td>POLR2E</td>\n",
       "      <td>4.054273</td>\n",
       "      <td>0.079496</td>\n",
       "      <td>5.029049e-05</td>\n",
       "      <td>4.368186e-01</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>22</td>\n",
       "      <td>25.0</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ZNF230</td>\n",
       "      <td>ZNF230</td>\n",
       "      <td>NaN</td>\n",
       "      <td>43.571832</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.590240e-06</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>C1QTNF4</td>\n",
       "      <td>C1QTNF4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.212809</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.062782e-05</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>7</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>MARK2</td>\n",
       "      <td>MARK2</td>\n",
       "      <td>3.881102</td>\n",
       "      <td>3.953367</td>\n",
       "      <td>1.039843e-04</td>\n",
       "      <td>1.185639e-04</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>12</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>PCDHA10</td>\n",
       "      <td>PCDHA10</td>\n",
       "      <td>3.832886</td>\n",
       "      <td>4.130926</td>\n",
       "      <td>1.266487e-04</td>\n",
       "      <td>1.267402e-04</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>64</td>\n",
       "      <td>67.0</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>PVR</td>\n",
       "      <td>PVR</td>\n",
       "      <td>-3.825266</td>\n",
       "      <td>-0.246338</td>\n",
       "      <td>1.306310e-04</td>\n",
       "      <td>2.794241e-02</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>82</td>\n",
       "      <td>89.0</td>\n",
       "      <td>89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ABCB9</td>\n",
       "      <td>ABCB9</td>\n",
       "      <td>-3.823876</td>\n",
       "      <td>-2.392964</td>\n",
       "      <td>1.313701e-04</td>\n",
       "      <td>3.315665e-04</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>29</td>\n",
       "      <td>30.0</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13839</th>\n",
       "      <td>BCL6</td>\n",
       "      <td>BCL6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13840</th>\n",
       "      <td>FAM129C</td>\n",
       "      <td>FAM129C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13841</th>\n",
       "      <td>PVRL2</td>\n",
       "      <td>PVRL2</td>\n",
       "      <td>-inf</td>\n",
       "      <td>-0.385584</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.270771e-01</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>137</td>\n",
       "      <td>149.0</td>\n",
       "      <td>151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13842</th>\n",
       "      <td>M6PR</td>\n",
       "      <td>M6PR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13843</th>\n",
       "      <td>SLC2A14</td>\n",
       "      <td>SLC2A14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13844</th>\n",
       "      <td>CLSTN3</td>\n",
       "      <td>CLSTN3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13845</th>\n",
       "      <td>CD163</td>\n",
       "      <td>CD163</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13846</th>\n",
       "      <td>RTN2</td>\n",
       "      <td>RTN2</td>\n",
       "      <td>-inf</td>\n",
       "      <td>-7.516555</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.305699e-04</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>19</td>\n",
       "      <td>23.0</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13847</th>\n",
       "      <td>C7</td>\n",
       "      <td>C7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13848</th>\n",
       "      <td>APOC2</td>\n",
       "      <td>APOC2</td>\n",
       "      <td>inf</td>\n",
       "      <td>3.803443</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.786886e-05</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>10</td>\n",
       "      <td>10.0</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13849</th>\n",
       "      <td>MFAP5</td>\n",
       "      <td>MFAP5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13850</th>\n",
       "      <td>PLEKHG6</td>\n",
       "      <td>PLEKHG6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13851</th>\n",
       "      <td>CD163L1</td>\n",
       "      <td>CD163L1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13852</th>\n",
       "      <td>TRIM40</td>\n",
       "      <td>TRIM40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13853</th>\n",
       "      <td>NECAP1</td>\n",
       "      <td>NECAP1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13854</th>\n",
       "      <td>ZNF180</td>\n",
       "      <td>ZNF180</td>\n",
       "      <td>inf</td>\n",
       "      <td>0.723927</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.089585e-02</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>135</td>\n",
       "      <td>145.0</td>\n",
       "      <td>153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13855</th>\n",
       "      <td>KLRG1</td>\n",
       "      <td>KLRG1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13856</th>\n",
       "      <td>VAC14</td>\n",
       "      <td>VAC14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13857</th>\n",
       "      <td>ANKRD30B</td>\n",
       "      <td>ANKRD30B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13858</th>\n",
       "      <td>RIMKLB</td>\n",
       "      <td>RIMKLB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13859</th>\n",
       "      <td>GABARAPL1</td>\n",
       "      <td>GABARAPL1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13860</th>\n",
       "      <td>C3AR1</td>\n",
       "      <td>C3AR1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13861</th>\n",
       "      <td>CNR2</td>\n",
       "      <td>CNR2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13862</th>\n",
       "      <td>A2M</td>\n",
       "      <td>A2M</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13863</th>\n",
       "      <td>FOXJ2</td>\n",
       "      <td>FOXJ2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13864</th>\n",
       "      <td>CD69</td>\n",
       "      <td>CD69</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13865</th>\n",
       "      <td>FAM90A1</td>\n",
       "      <td>FAM90A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13866</th>\n",
       "      <td>SMIM15</td>\n",
       "      <td>SMIM15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13867</th>\n",
       "      <td>VAC14</td>\n",
       "      <td>VAC14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13868</th>\n",
       "      <td>CLEC4A</td>\n",
       "      <td>CLEC4A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>1.000000e-10</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13869 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene  gene_name    zscore  effect_size        pvalue  \\\n",
       "0       CEACAM19   CEACAM19  7.570289     0.494302  3.723957e-14   \n",
       "1          EXTL3      EXTL3       NaN   -23.451963           NaN   \n",
       "2         CLPTM1     CLPTM1       NaN    -1.099350           NaN   \n",
       "3            CLU        CLU  6.973488     0.043537  3.091788e-12   \n",
       "4           BCAM       BCAM       NaN     2.538550           NaN   \n",
       "5        PPP1R37    PPP1R37  6.430535    15.715981  1.271553e-10   \n",
       "6          EPHA1      EPHA1  6.103094     1.449575  1.040343e-09   \n",
       "7          CASP2      CASP2       NaN     3.117632           NaN   \n",
       "8          MTCH2      MTCH2 -5.462552    -2.070170  4.693376e-08   \n",
       "9          CD2AP      CD2AP  4.860381     0.291114  1.171603e-06   \n",
       "10         SYTL2      SYTL2 -4.846171    -0.574747  1.258672e-06   \n",
       "11       FAM180B    FAM180B       NaN   -24.396055           NaN   \n",
       "12        PTGDR2     PTGDR2 -4.586619    -1.197283  4.504807e-06   \n",
       "13        NDUFS3     NDUFS3       NaN    14.895390           NaN   \n",
       "14         WDR55      WDR55  4.464619     1.377777  8.021139e-06   \n",
       "15         AIMP2      AIMP2 -4.463330    -2.005129  8.069562e-06   \n",
       "16          CD14       CD14  4.414933     0.492409  1.010413e-05   \n",
       "17           ZP1        ZP1 -4.382206    -2.423554  1.174834e-05   \n",
       "18          MAFK       MAFK  4.339047     0.301414  1.431018e-05   \n",
       "19         RAPSN      RAPSN       NaN   -83.548762           NaN   \n",
       "20         HBEGF      HBEGF       NaN    -9.826707           NaN   \n",
       "21          DMWD       DMWD  4.148808     0.431981  3.342103e-05   \n",
       "22           MDK        MDK       NaN    -0.963345           NaN   \n",
       "23        POLR2E     POLR2E  4.054273     0.079496  5.029049e-05   \n",
       "24        ZNF230     ZNF230       NaN    43.571832           NaN   \n",
       "25       C1QTNF4    C1QTNF4       NaN     8.212809           NaN   \n",
       "26         MARK2      MARK2  3.881102     3.953367  1.039843e-04   \n",
       "27       PCDHA10    PCDHA10  3.832886     4.130926  1.266487e-04   \n",
       "28           PVR        PVR -3.825266    -0.246338  1.306310e-04   \n",
       "29         ABCB9      ABCB9 -3.823876    -2.392964  1.313701e-04   \n",
       "...          ...        ...       ...          ...           ...   \n",
       "13839       BCL6       BCL6       NaN          NaN           NaN   \n",
       "13840    FAM129C    FAM129C       NaN          NaN           NaN   \n",
       "13841      PVRL2      PVRL2      -inf    -0.385584           NaN   \n",
       "13842       M6PR       M6PR       NaN          NaN           NaN   \n",
       "13843    SLC2A14    SLC2A14       NaN          NaN           NaN   \n",
       "13844     CLSTN3     CLSTN3       NaN          NaN           NaN   \n",
       "13845      CD163      CD163       NaN          NaN           NaN   \n",
       "13846       RTN2       RTN2      -inf    -7.516555           NaN   \n",
       "13847         C7         C7       NaN          NaN           NaN   \n",
       "13848      APOC2      APOC2       inf     3.803443           NaN   \n",
       "13849      MFAP5      MFAP5       NaN          NaN           NaN   \n",
       "13850    PLEKHG6    PLEKHG6       NaN          NaN           NaN   \n",
       "13851    CD163L1    CD163L1       NaN          NaN           NaN   \n",
       "13852     TRIM40     TRIM40       NaN          NaN           NaN   \n",
       "13853     NECAP1     NECAP1       NaN          NaN           NaN   \n",
       "13854     ZNF180     ZNF180       inf     0.723927           NaN   \n",
       "13855      KLRG1      KLRG1       NaN          NaN           NaN   \n",
       "13856      VAC14      VAC14       NaN          NaN           NaN   \n",
       "13857   ANKRD30B   ANKRD30B       NaN          NaN           NaN   \n",
       "13858     RIMKLB     RIMKLB       NaN          NaN           NaN   \n",
       "13859  GABARAPL1  GABARAPL1       NaN          NaN           NaN   \n",
       "13860      C3AR1      C3AR1       NaN          NaN           NaN   \n",
       "13861       CNR2       CNR2       NaN          NaN           NaN   \n",
       "13862        A2M        A2M       NaN          NaN           NaN   \n",
       "13863      FOXJ2      FOXJ2       NaN          NaN           NaN   \n",
       "13864       CD69       CD69       NaN          NaN           NaN   \n",
       "13865    FAM90A1    FAM90A1       NaN          NaN           NaN   \n",
       "13866     SMIM15     SMIM15       NaN          NaN           NaN   \n",
       "13867      VAC14      VAC14       NaN          NaN           NaN   \n",
       "13868     CLEC4A     CLEC4A       NaN          NaN           NaN   \n",
       "\n",
       "              var_g  pred_perf_r2  pred_perf_pval  pred_perf_qval  \\\n",
       "0      3.791226e-02           0.5    1.000000e-10    1.000000e-10   \n",
       "1      1.264877e-05           0.5    1.000000e-10    1.000000e-10   \n",
       "2      5.464133e-03           0.5    1.000000e-10    1.000000e-10   \n",
       "3      3.343004e+00           0.5    1.000000e-10    1.000000e-10   \n",
       "4      9.117730e-04           0.5    1.000000e-10    1.000000e-10   \n",
       "5      3.665028e-05           0.5    1.000000e-10    1.000000e-10   \n",
       "6      2.392429e-03           0.5    1.000000e-10    1.000000e-10   \n",
       "7      4.083994e-04           0.5    1.000000e-10    1.000000e-10   \n",
       "8      9.155414e-04           0.5    1.000000e-10    1.000000e-10   \n",
       "9      3.277740e-02           0.5    1.000000e-10    1.000000e-10   \n",
       "10     8.906129e-03           0.5    1.000000e-10    1.000000e-10   \n",
       "11     4.581500e-06           0.5    1.000000e-10    1.000000e-10   \n",
       "12     1.779324e-03           0.5    1.000000e-10    1.000000e-10   \n",
       "13     1.271488e-05           0.5    1.000000e-10    1.000000e-10   \n",
       "14     1.333930e-03           0.5    1.000000e-10    1.000000e-10   \n",
       "15     1.141239e-03           0.5    1.000000e-10    1.000000e-10   \n",
       "16     5.269745e-03           0.5    1.000000e-10    1.000000e-10   \n",
       "17     4.078626e-04           0.5    1.000000e-10    1.000000e-10   \n",
       "18     2.929442e-02           0.5    1.000000e-10    1.000000e-10   \n",
       "19     3.574258e-07           0.5    1.000000e-10    1.000000e-10   \n",
       "20     2.437816e-05           0.5    1.000000e-10    1.000000e-10   \n",
       "21     1.295833e-02           0.5    1.000000e-10    1.000000e-10   \n",
       "22     2.461208e-03           0.5    1.000000e-10    1.000000e-10   \n",
       "23     4.368186e-01           0.5    1.000000e-10    1.000000e-10   \n",
       "24     1.590240e-06           0.5    1.000000e-10    1.000000e-10   \n",
       "25     3.062782e-05           0.5    1.000000e-10    1.000000e-10   \n",
       "26     1.185639e-04           0.5    1.000000e-10    1.000000e-10   \n",
       "27     1.267402e-04           0.5    1.000000e-10    1.000000e-10   \n",
       "28     2.794241e-02           0.5    1.000000e-10    1.000000e-10   \n",
       "29     3.315665e-04           0.5    1.000000e-10    1.000000e-10   \n",
       "...             ...           ...             ...             ...   \n",
       "13839           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "13840           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "13841  1.270771e-01           0.5    1.000000e-10    1.000000e-10   \n",
       "13842           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "13843           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "13844           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "13845           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "13846  1.305699e-04           0.5    1.000000e-10    1.000000e-10   \n",
       "13847           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "13848  3.786886e-05           0.5    1.000000e-10    1.000000e-10   \n",
       "13849           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "13850           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "13851           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "13852           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "13853           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "13854  1.089585e-02           0.5    1.000000e-10    1.000000e-10   \n",
       "13855           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "13856           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "13857           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "13858           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "13859           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "13860           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "13861           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "13862           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "13863           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "13864           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "13865           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "13866           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "13867           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "13868           NaN           0.5    1.000000e-10    1.000000e-10   \n",
       "\n",
       "       n_snps_used  n_snps_in_cov  n_snps_in_model  \n",
       "0              143          162.0              163  \n",
       "1                4            4.0                4  \n",
       "2               21           22.0               22  \n",
       "3               21           21.0               21  \n",
       "4                1            1.0                1  \n",
       "5                5            5.0                6  \n",
       "6               26           27.0               27  \n",
       "7               17           20.0               20  \n",
       "8               36           37.0               38  \n",
       "9               37           37.0               37  \n",
       "10              36           38.0               39  \n",
       "11               2            2.0                2  \n",
       "12              12           12.0               12  \n",
       "13               1            1.0                1  \n",
       "14              55           60.0               63  \n",
       "15              17           22.0               22  \n",
       "16              18           19.0               19  \n",
       "17              43           47.0               51  \n",
       "18              34           47.0               48  \n",
       "19               2            3.0                3  \n",
       "20               3            7.0                7  \n",
       "21              55           57.0               57  \n",
       "22               4            4.0                4  \n",
       "23              22           25.0               26  \n",
       "24               1            1.0                1  \n",
       "25               7            7.0                7  \n",
       "26              12           12.0               12  \n",
       "27              64           67.0               69  \n",
       "28              82           89.0               89  \n",
       "29              29           30.0               30  \n",
       "...            ...            ...              ...  \n",
       "13839            0            1.0                2  \n",
       "13840            0            NaN                1  \n",
       "13841          137          149.0              151  \n",
       "13842            0            1.0               55  \n",
       "13843            0            NaN              127  \n",
       "13844            0            NaN               13  \n",
       "13845            0            NaN               43  \n",
       "13846           19           23.0               24  \n",
       "13847            0            NaN                1  \n",
       "13848           10           10.0               10  \n",
       "13849            0            NaN               27  \n",
       "13850            0            NaN                1  \n",
       "13851            0            NaN                5  \n",
       "13852            0           45.0               46  \n",
       "13853            0            NaN                3  \n",
       "13854          135          145.0              153  \n",
       "13855            0            NaN               31  \n",
       "13856            0            1.0                3  \n",
       "13857            0            1.0                2  \n",
       "13858            0            NaN                4  \n",
       "13859            0            NaN                1  \n",
       "13860            0            NaN               64  \n",
       "13861            0            1.0                2  \n",
       "13862            0            NaN                7  \n",
       "13863            0            NaN               47  \n",
       "13864            0            NaN                3  \n",
       "13865            0            NaN               65  \n",
       "13866            0            NaN                1  \n",
       "13867            0            1.0                3  \n",
       "13868            0            NaN                3  \n",
       "\n",
       "[13869 rows x 12 columns]"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = zscore_dict[nam]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0         CEACAM19\n",
       "1            EXTL3\n",
       "2           CLPTM1\n",
       "3              CLU\n",
       "4             BCAM\n",
       "5          PPP1R37\n",
       "6            EPHA1\n",
       "7            CASP2\n",
       "8            MTCH2\n",
       "9            CD2AP\n",
       "10           SYTL2\n",
       "11         FAM180B\n",
       "12          PTGDR2\n",
       "13          NDUFS3\n",
       "14           WDR55\n",
       "15           AIMP2\n",
       "16            CD14\n",
       "17             ZP1\n",
       "18            MAFK\n",
       "19           RAPSN\n",
       "20           HBEGF\n",
       "21            DMWD\n",
       "22             MDK\n",
       "23          POLR2E\n",
       "24          ZNF230\n",
       "25         C1QTNF4\n",
       "26           MARK2\n",
       "27         PCDHA10\n",
       "28             PVR\n",
       "29           ABCB9\n",
       "           ...    \n",
       "13839         BCL6\n",
       "13840      FAM129C\n",
       "13841        PVRL2\n",
       "13842         M6PR\n",
       "13843      SLC2A14\n",
       "13844       CLSTN3\n",
       "13845        CD163\n",
       "13846         RTN2\n",
       "13847           C7\n",
       "13848        APOC2\n",
       "13849        MFAP5\n",
       "13850      PLEKHG6\n",
       "13851      CD163L1\n",
       "13852       TRIM40\n",
       "13853       NECAP1\n",
       "13854       ZNF180\n",
       "13855        KLRG1\n",
       "13856        VAC14\n",
       "13857     ANKRD30B\n",
       "13858       RIMKLB\n",
       "13859    GABARAPL1\n",
       "13860        C3AR1\n",
       "13861         CNR2\n",
       "13862          A2M\n",
       "13863        FOXJ2\n",
       "13864         CD69\n",
       "13865      FAM90A1\n",
       "13866       SMIM15\n",
       "13867        VAC14\n",
       "13868       CLEC4A\n",
       "Name: gene, Length: 13869, dtype: object"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"gene\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        False\n",
       "1        False\n",
       "2        False\n",
       "3        False\n",
       "4        False\n",
       "5        False\n",
       "6        False\n",
       "7        False\n",
       "8        False\n",
       "9        False\n",
       "10       False\n",
       "11       False\n",
       "12       False\n",
       "13       False\n",
       "14       False\n",
       "15       False\n",
       "16       False\n",
       "17       False\n",
       "18       False\n",
       "19       False\n",
       "20       False\n",
       "21       False\n",
       "22       False\n",
       "23       False\n",
       "24       False\n",
       "25       False\n",
       "26       False\n",
       "27       False\n",
       "28       False\n",
       "29       False\n",
       "         ...  \n",
       "13839    False\n",
       "13840    False\n",
       "13841    False\n",
       "13842    False\n",
       "13843    False\n",
       "13844    False\n",
       "13845    False\n",
       "13846    False\n",
       "13847    False\n",
       "13848    False\n",
       "13849    False\n",
       "13850    False\n",
       "13851    False\n",
       "13852    False\n",
       "13853    False\n",
       "13854    False\n",
       "13855    False\n",
       "13856    False\n",
       "13857    False\n",
       "13858    False\n",
       "13859    False\n",
       "13860    False\n",
       "13861    False\n",
       "13862    False\n",
       "13863    False\n",
       "13864    False\n",
       "13865    False\n",
       "13866    False\n",
       "13867    False\n",
       "13868    False\n",
       "Name: gene, Length: 13869, dtype: bool"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "index = zscore_dict[nam][\"gene\"] == gene\n",
    "index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[  1.00000000e+00,   1.35881564e-06,   7.86645134e-05],\n",
       "       [  1.35881564e-06,   6.11736973e-07,   1.10037953e-06],\n",
       "       [  7.86645134e-05,   1.10037953e-06,   1.90250017e-05]])"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cov_gene.shape\n",
    "cov_gene[:3,:3]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0371856561492391e-10"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cov_gene[1,5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0371856561492392e-10"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cov_gene[5,1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.allclose(cov_gene,cov_gene.T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7.4587912223294393e-08"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cov_gene[8,2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7.4587912223294393e-08"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cov_gene[2,8]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### rpy test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import rpy2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def r_requirement():\n",
    "    utils = rpackages.importr('utils')\n",
    "    # select a mirror for R packages\n",
    "    utils.chooseCRANmirror(ind=1) # select the first mirror in the list\n",
    "    \n",
    "    # R package names\n",
    "    names_to_install = ['GBJ']\n",
    "\n",
    "    # install R packages\n",
    "    from rpy2.robjects.vectors import StrVector  \n",
    "    if len(names_to_install) > 0:\n",
    "        utils.install_packages(StrVector(names_to_install))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/rinterface/__init__.py:186: RRuntimeWarning: also installing the dependencies ‘Rcpp’, ‘mvtnorm’, ‘SKAT’\n",
      "\n",
      "\n",
      "  warnings.warn(x, RRuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "r_requirement()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "ename": "RRuntimeError",
     "evalue": "Error in loadNamespace(name) : there is no package called ‘GBJ’\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRRuntimeError\u001b[0m                             Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-134-98186b5c4b42>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mimportr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'GBJ'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/Users/jerome/anaconda3/envs/metaxcan/lib/python2.7/site-packages/rpy2/robjects/packages.pyc\u001b[0m in \u001b[0;36mimportr\u001b[0;34m(name, lib_loc, robject_translations, signature_translation, suppress_messages, on_conflict, symbol_r2python, symbol_check_after, data)\u001b[0m\n\u001b[1;32m    451\u001b[0m     if _package_has_namespace(rname, \n\u001b[1;32m    452\u001b[0m                               _system_file(package = rname)):\n\u001b[0;32m--> 453\u001b[0;31m         \u001b[0menv\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_get_namespace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    454\u001b[0m         \u001b[0mversion\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_get_namespace_version\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    455\u001b[0m         \u001b[0mexported_names\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_get_namespace_exports\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mRRuntimeError\u001b[0m: Error in loadNamespace(name) : there is no package called ‘GBJ’\n"
     ]
    }
   ],
   "source": [
    "importr('GBJ')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
